### PR TITLE
(logs): Query range v3

### PR DESCRIPTION
**Pull request by nityanandagohain**
Part of #2250 

Examples are present in postman`Query Range v3 logs` folder

### CODE REVIEW

**Comment by nityanandagohain** [comment]
In the `listPanelType` implementation for `SelectColumns` is left in this PR, I am figuring out that part. As of now all values are returned.

File: pkg/query-service/model/v3/v3.go [hunk 1]
```diff
@@ -264,6 +264,9 @@ func (a AttributeKey) Validate() error {
 	switch a.DataType {
 	case AttributeKeyDataTypeBool, AttributeKeyDataTypeInt64, AttributeKeyDataTypeFloat64, AttributeKeyDataTypeString:
 		break
+	case "":
+		// allow empty values to pass through
+		break
```
Comment (srikanthccv) [inline pkg/query-service/model/v3/v3.go:269]: I would add a new type Unspecified, and keep part of the above case statement.

File: pkg/query-service/app/http_handler.go [hunk 2]
```diff
@@ -2617,11 +2729,13 @@ func (aH *APIHandler) queryRangeV3(queryRangeParams *v3.QueryRangeParamsV3, w ht
...
+	ctx := context.Background()
+
 	if apiErrorObj != nil {
 		zap.S().Errorf(apiErrorObj.Err.Error())
 		RespondError(w, apiErrorObj, nil)
 		return
 	}
 
-	aH.queryRangeV3(queryRangeParams, w, r)
+	aH.queryRangeV3(ctx, queryRangeParams, w, r)
```
Comment (srikanthccv) [inline pkg/query-service/app/http_handler.go:2740]:
```suggestion
	aH.queryRangeV3(r.Context(), queryRangeParams, w, r)
```

File: pkg/query-service/app/logs/v3/query_builder.go [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+		selectLabels = ""
+	} else {
+		for _, tag := range groupBy {
+			columnName, err := getClickhouseColumnName(tag, fields)
+			if err != nil {
+				return "", err
+			}
+			selectLabels += fmt.Sprintf(", %s as %s", columnName, tag.Key)
+		}
+	}
```
Comment (srikanthccv) [inline pkg/query-service/app/logs/v3/query_builder.go:109]: The `rate` in metrics is a bit different from the `rate` in logs, You could have the rate of logs by service, host, so you need to handle that. You seem to skip that for `rate` here?

File: pkg/query-service/app/logs/v3/query_builder.go [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+}
+
+func buildLogsQuery(start, end, step int64, mq *v3.BuilderQuery, tableName string, fields map[string]v3.AttributeKey) (string, error) {
+
+	filterSubQuery, err := buildLogsTimeSeriesFilterQuery(mq.Filters, fields)
+	if err != nil {
+		return "", err
+	}
+
+	samplesTableTimeFilter := fmt.Sprintf("(timestamp >= %d AND timestamp <= %d)", start, end)
```
Comment (srikanthccv) [inline pkg/query-service/app/logs/v3/query_builder.go:150]:
```suggestion
	timeFilter := fmt.Sprintf("(timestamp >= %d AND timestamp <= %d)", start, end)
```

Check the precision

File: pkg/query-service/app/logs/v3/query_builder.go [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+		op := fmt.Sprintf("quantile(%v)(%s)", aggregateOperatorToPercentile[mq.AggregateOperator], aggregationKey)
+		query := fmt.Sprintf(queryTmpl, step, op, filterSubQuery, groupBy, orderBy)
+		return query, nil
+	case v3.AggregateOperatorAvg, v3.AggregateOperatorSum, v3.AggregateOperatorMin, v3.AggregateOperatorMax:
+		op := fmt.Sprintf("%s(%s)", aggregateOperatorToSQLFunc[mq.AggregateOperator], aggregationKey)
+		query := fmt.Sprintf(queryTmpl, step, op, filterSubQuery, groupBy, orderBy)
+		return query, nil
+	case v3.AggregateOpeatorCount:
+		op := "toFloat64(count(*))"
+		query := fmt.Sprintf(queryTmpl, step, op, filterSubQuery, groupBy, orderBy)
```
Comment (srikanthccv) [inline pkg/query-service/app/logs/v3/query_builder.go:208]: If they provide an aggregate key for `count`, they mean that it exists. Is that honoured? For instance, count of `passthrough.this.codepath` by service. Here I am trying to count the number of records seen for each interval with `passthrough.this.codepath`. It appears to me that currently, it counts all regardless of the existence of `passthrough.this.codepath` which is not the expected behaviour.

File: pkg/query-service/app/logs/v3/query_builder.go [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+				orderBy = append(orderBy, fmt.Sprintf("%s %s", item.ColumnName, item.Order))
+				break
+			}
+		}
+		if !found {
+			orderBy = append(orderBy, fmt.Sprintf("%s ASC", tag))
+		}
+	}
+	return strings.Join(orderBy, ",")
+}
```
Comment (srikanthccv) [inline pkg/query-service/app/logs/v3/query_builder.go:256]: The `Order By` option will not be there for metrics but will be commonly used for logs/traces for the end user. There is an option to order by aggregate value. Is that considered here?

File: pkg/query-service/app/logs/v3/query_builder.go [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+	}
+	return fmt.Sprintf("%s LIMIT %d", query, limit)
+}
+
+func addOffsetToQuery(query string, offset uint64) string {
+	return fmt.Sprintf("%s OFFSET %d", query, offset)
+}
+
+func PrepareLogsQuery(start, end, step int64, queryType v3.QueryType, panelType v3.PanelType, mq *v3.BuilderQuery, fields map[string]v3.AttributeKey) (string, error) {
+	query, err := buildLogsQuery(start, end, step, mq, constants.SIGNOZ_TIMESERIES_TABLENAME, fields)
```
Comment (srikanthccv) [inline pkg/query-service/app/logs/v3/query_builder.go:311]:
```suggestion
	query, err := buildLogsQuery(start, end, step, mq, fields)
```

File: pkg/query-service/app/logs/v3/query_builder.go (hunk 4 above)
Comment (nityanandagohain) [reply pkg/query-service/app/logs/v3/query_builder.go:150]: Ahh so in my payload I was sending Epochnano that's why missed this. Keeping it epoch milliseconds for all.

Comment (nityanandagohain) [reply pkg/query-service/app/logs/v3/query_builder.go]: Added a function to which will convert it to nano if the input is in milli/micro.

File: pkg/query-service/app/logs/v3/query_builder.go (hunk 6 above)
Comment (nityanandagohain) [reply pkg/query-service/app/logs/v3/query_builder.go:278]: Have added support for it

File: pkg/query-service/app/http_handler.go [hunk 8]
```diff
@@ -2572,19 +2620,83 @@ func (aH *APIHandler) execPromQueries(ctx context.Context, metricsQueryRangePara
...
+			getType := func(t string) (v3.AttributeKeyType, bool) {
+				if t == "attributes" {
+					return v3.AttributeKeyTypeTag, false
+				} else if t == "resources" {
+					return v3.AttributeKeyTypeResource, false
+				}
+				return "", true
+			}
+
+			for _, v := range fields.Selected {
```
Comment (srikanthccv) [inline pkg/query-service/app/http_handler.go:2642]: There is a `v` here and in the outer loop, making it hard to follow. Please use distinct/descriptive names to make it easier for the reader.

File: pkg/query-service/app/logs/v3/query_builder.go [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+	v3.AggregateOperatorP50:         0.50,
+	v3.AggregateOperatorP75:         0.75,
+	v3.AggregateOperatorP90:         0.90,
+	v3.AggregateOperatorP95:         0.95,
+	v3.AggregateOperatorP99:         0.99,
+	v3.AggregateOperatorHistQuant50: 0.50,
+	v3.AggregateOperatorHistQuant75: 0.75,
+	v3.AggregateOperatorHistQuant90: 0.90,
+	v3.AggregateOperatorHistQuant95: 0.95,
+	v3.AggregateOperatorHistQuant99: 0.99,
```
Comment (srikanthccv) [inline pkg/query-service/app/logs/v3/query_builder.go:27]:
```suggestion
```

File: pkg/query-service/app/logs/v3/query_builder.go [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+	v3.FilterOperatorLessThanOrEq:    {},
+	v3.FilterOperatorGreaterThan:     {},
+	v3.FilterOperatorGreaterThanOrEq: {},
+	v3.FilterOperatorLike:            {},
+	v3.FilterOperatorNotLike:         {},
+	v3.FilterOperatorRegex:           {},
+	v3.FilterOperatorNotRegex:        {},
+	v3.FilterOperatorIn:              {},
+	v3.FilterOperatorNotIn:           {},
+	// (todo) check contains/not contains/exists/not exists
```
Comment (srikanthccv) [inline pkg/query-service/app/logs/v3/query_builder.go:54]: These needs to be supported as well.

File: pkg/query-service/app/logs/v3/query_builder.go [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+		}
+
+		columnDataType := "string"
+		if key.DataType == v3.AttributeKeyDataTypeFloat64 {
+			columnDataType = "float64"
+		}
+		if key.DataType == v3.AttributeKeyDataTypeInt64 {
+			columnDataType = "int64"
+		}
+		clickhouseColumn = fmt.Sprintf("%s_%s_value[indexOf(%s_%s_key, '%s')]", columnType, columnDataType, columnType, columnDataType, key.Key)
```
Comment (srikanthccv) [inline pkg/query-service/app/logs/v3/query_builder.go:92]: Are booleans skipped somewhere? Otherwise, they would result in the wrong query for the query range?

File: pkg/query-service/app/logs/v3/query_builder.go [hunk 12]
```diff
@@ -0,0 +1,349 @@
...
+
+			// generate the key
+			columnName, err := getClickhouseColumnName(item.Key, fields)
+			if err != nil {
+				return "", err
+			}
+			fmtVal := utils.ClickHouseFormattedValue(toFormat)
+
+			if _, ok := logsOperatorsEnabled[op]; ok {
+				conditions = append(conditions, fmt.Sprintf("%s %s %s", columnName, op, fmtVal))
```
Comment (srikanthccv) [inline pkg/query-service/app/logs/v3/query_builder.go:130]: This fails for operators NIN, NREGEX, NCONTAINS etc.

File: pkg/query-service/app/logs/v3/query_builder.go [hunk 13]
```diff
@@ -0,0 +1,349 @@
...
+		for _, item := range items {
+			if item.ColumnName == tag {
+				found = true
+				orderBy = append(orderBy, fmt.Sprintf("%s %s", item.ColumnName, item.Order))
+				break
+			}
+		}
+		if !found {
+			orderBy = append(orderBy, fmt.Sprintf("%s ASC", tag))
+		}
```
Comment (srikanthccv) [inline pkg/query-service/app/logs/v3/query_builder.go:268]: Does this not conflict with the aggregate value order by? It wouldn't appear in the group by tags?

File: pkg/query-service/app/logs/v3/query_builder.go (hunk 5 above)
Comment (nityanandagohain) [reply pkg/query-service/app/logs/v3/query_builder.go:245]: Have added a `has` clause.

File: pkg/query-service/app/logs/v3/query_builder.go (hunk 11 above)
Comment (nityanandagohain) [reply pkg/query-service/app/logs/v3/query_builder.go:104]: Right, as of now in the OTEL collector bool data is stored as a string https://github.com/SigNoz/signoz-otel-collector/blob/7338768fa6b2c61aa8c46f45c9c7de700e2795b8/exporter/clickhouselogsexporter/migrations/000001_init_db.up.sql#L14 

Have added comments in the code for that.

File: pkg/query-service/app/logs/v3/query_builder.go (hunk 13 above)
Comment (nityanandagohain) [reply pkg/query-service/app/logs/v3/query_builder.go:286]: That logic is added below and outside this loop, so it won't conflict.

File: pkg/query-service/app/logs/v3/query_builder.go (hunk 12 above)
Comment (nityanandagohain) [reply pkg/query-service/app/logs/v3/query_builder.go]: I have updated the test for `timeSeriesFilterQueryData` for `in` / `nin` operators, can you check once.

File: pkg/query-service/app/logs/v3/query_builder.go (hunk 13 above)
Comment (srikanthccv) [reply pkg/query-service/app/logs/v3/query_builder.go:286]: Can you add a test with `SigNozOrderByValue` present in the order by of the query payload? I hope that makes it clear.

File: pkg/query-service/app/logs/v3/query_builder.go (hunk 10 above)
Comment (nityanandagohain) [reply pkg/query-service/app/logs/v3/query_builder.go:49]: I didn't find any difference between `contains/not contains` and `like/not like`. Even in metrics implementation, it's the same.

In logs previously we converted contains to -> `ilike`. I am not really sure what should we do.

Comment (srikanthccv) [reply pkg/query-service/app/logs/v3/query_builder.go:49]: I didn't have the `contains/not contains` earlier, so I imagined this was important for logs/traces. The difference, I believe, is `contains` is regular text, but the `like/not like` is an expression in the format which includes `%`, `_` for matching. For example, `status_code` like `2__` matches the `2xx`, but that would differ from the `status_code` contains `2`.

Comment (nityanandagohain) [reply pkg/query-service/app/logs/v3/query_builder.go]: Have added contains,ncontains,exists,nexists

File: pkg/query-service/app/logs/v3/query_builder.go (hunk 13 above)
Comment (nityanandagohain) [reply pkg/query-service/app/logs/v3/query_builder.go:314]: Have added tests for order by, please check one.

File: pkg/query-service/app/logs/v3/query_builder_test.go [hunk 14]
```diff
@@ -0,0 +1,699 @@
...
+		GroupByTags:       []v3.AttributeKey{{Key: "host", IsColumn: true}},
+		SelectLabels:      ", host as host",
+	},
+}
+
+func TestGetSelectLabels(t *testing.T) {
+	for _, tt := range testGetSelectLabelsData {
+		Convey("testGetSelectLabelsData", t, func() {
+			selectLabels, err := getSelectLabels(tt.AggregateOperator, tt.GroupByTags, map[string]v3.AttributeKey{})
+			fmt.Println(selectLabels)
```
Comment (srikanthccv) [inline pkg/query-service/app/logs/v3/query_builder_test.go:93]:
```suggestion
```

File: pkg/query-service/app/logs/v3/query_builder_test.go [hunk 15]
```diff
@@ -0,0 +1,699 @@
...
+	},
+}
+
+func TestBuildLogsTimeSeriesFilterQuery(t *testing.T) {
+	for _, tt := range timeSeriesFilterQueryData {
+		Convey("TestBuildLogsTimeSeriesFilterQuery", t, func() {
+			query, err := buildLogsTimeSeriesFilterQuery(tt.FilterSet, map[string]v3.AttributeKey{})
+			So(err, ShouldBeNil)
+			So(query, ShouldEqual, tt.ExpectedFilter)
+			fmt.Println(query)
```
Comment (srikanthccv) [inline pkg/query-service/app/logs/v3/query_builder_test.go:185]:
```suggestion
```

File: pkg/query-service/app/logs/v3/query_builder_test.go [hunk 16]
```diff
@@ -0,0 +1,699 @@
...
+		TableName:     "logs",
+		ExpectedQuery: "SELECT toStartOfInterval(fromUnixTimestamp64Nano(timestamp), INTERVAL 60 SECOND) AS ts, toFloat64(count(distinct(attributes_string_value[indexOf(attributes_string_key, 'name')]))) as value from signoz_logs.distributed_logs where (timestamp >= 1680066360726210000 AND timestamp <= 1680066458000000000) AND attributes_string_value[indexOf(attributes_string_key, 'method')] = 'GET' group by ts having value > 10 order by ts",
+	},
+}
+
+func TestBuildLogsQuery(t *testing.T) {
+	for _, tt := range testBuildLogsQueryData {
+		Convey("TestBuildLogsQuery", t, func() {
+			query, err := buildLogsQuery(tt.Start, tt.End, tt.Step, tt.BuilderQuery, map[string]v3.AttributeKey{})
+			fmt.Println(query)
```
Comment (srikanthccv) [inline pkg/query-service/app/logs/v3/query_builder_test.go:594]:
```suggestion
```

**Review by srikanthccv — DISMISSED** [review]

**Review by srikanthccv — APPROVED** [review]

**Review by makeavish — DISMISSED** [review]
LGTM!
Please create issues for the remaining items.

File: pkg/query-service/app/logs/v3/query_builder.go [hunk 17]
```diff
@@ -0,0 +1,397 @@
+package v3
+
+import (
+	"fmt"
```
Comment (srikanthccv) [inline pkg/query-service/app/logs/v3/query_builder.go:4]:
```suggestion
```

**Comment by srikanthccv** [comment]
Add the commit suggestion to remove the unused import, which was causing the tests to fail.

**Review by srikanthccv — APPROVED** [review]

**Review by srikanthccv — APPROVED** [review]
